### PR TITLE
PMM-7820 add dbaas toggle

### PR DESCRIPTION
**Pull request by jprukner**
[PMM-7820](https://jira.percona.com/browse/PMM-7820)

To avoid reconnecting to dbaas-controller which may not be running, I came up with this rather complex solution for dbaas client. Simply passing the api address and creating dbaas client for every request is not an option because we need to be able to pass mocked dbaas client to dbaas services.

- [x] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1688

- [ ] Links to other linked pull requests (optional).
---
- [ ] Tests passed.
- [ ] Feature build pass.
- [ ] (Re)requested review.
- [ ] Fix conflicts with target branch.
- [ ] Update API dependency.
- [ ] Update jira ticket description if needed.
- [ ] Attach screenshots/console output to confirm new behavior to jira ticket, if applicable.

### CODE REVIEW

File: services/supervisord/supervisord.go [hunk 1]
```diff
@@ -205,8 +205,11 @@ func (s *Service) ForceCheckUpdates(ctx context.Context) error {
 	return s.pmmUpdateCheck.check(ctx)
 }
 
-func (s *Service) subscribe(program string, eventTypes ...eventType) chan *event {
-	ch := make(chan *event, 1)
+// Subscribe returns a channel that is fed by one of eventTypes. When when event
+// is sent to the channle, the channel is closed. User of this method should call
```
Comment (palash25) [inline services/supervisord/supervisord.go:209]: `channle` -> `channel`

**Review by palash25 — CHANGES_REQUESTED** [review]

File: main.go [hunk 2]
```diff
@@ -737,7 +711,22 @@ func main() {
 		l.Fatalf("Failed to get settings: %+v.", err)
 	}
 
-	dbaasControllerClient := getDBaaSControllerClient(ctx, *dbaasControllerAPIAddrF, settings)
+	dbaasClient := dbaas.NewClient(*dbaasControllerAPIAddrF, supervisord)
+	if settings.DBaaS.Enabled {
+		l.Debug("creating a new dbaas client in main - dbaas is enabled")
```
[the remark this answers is not in the document]
Comment (jprukner) [reply main.go]: Done

File: main.go [hunk 3]
```diff
@@ -795,21 +784,22 @@ func main() {
...
-			jobsService:           jobsService,
-			versionServiceClient:  versionService,
+			db:             db,
+			vmdb:           vmdb,
+			server:         server,
+			agentsRegistry: agentsRegistry,
+			grafanaClient:  grafanaClient,
+			checksService:  checksService,
+			dbaasClient:    dbaasClient,
+
```
[the remark this answers is not in the document]
Comment (jprukner) [reply main.go]: Removed

File: services/supervisord/supervisord.go (hunk 1 above)
Comment (jprukner) [reply services/supervisord/supervisord.go]: Changed

File: services/dbaas/dbaas_client.go [hunk 4]
```diff
@@ -19,102 +19,476 @@ package dbaas
...
+				var ctx context.Context
+				ctx, localCancel := context.WithTimeout(context.Background(), time.Second*20)
+				select {
+				case c.connectCh <- connectRequest{ctx: ctx}:
+					cancel = localCancel
+				default:
+					localCancel()
+					c.l.Warn("Request to connect to dbaas-controller API was already sent.")
+				}
+			case supervisord.Stopping, supervisord.ExitedUnexpected, supervisord.ExitedExpected:
```
Comment (BupycHuk) [inline services/dbaas/dbaas_client.go:181]: I would suggest to disconnect only if `Disconnect()` is called and call it if `DBaaSEnabled` changed to false.
The same about `Connect`. No need to parse supervisord events.
If dbaas-controller fails unexpected, supervisor will restart it, so dbaasClient should retry to connect to dbaas-controller.
If dbaas-controller fails continously then user will see high CPU usage and will be able to disable dbaas by themself.

File: services/dbaas/dbaas_client.go [hunk 5]
```diff
@@ -19,102 +19,476 @@ package dbaas
...
+	if resp.err != nil {
+		return nil, resp.err
+	}
+	return resp.out.(*controllerv1beta1.CreateXtraDBClusterResponse), resp.err
 }
 
 // UpdateXtraDBCluster updates existing XtraDB cluster.
 func (c *Client) UpdateXtraDBCluster(ctx context.Context, in *controllerv1beta1.UpdateXtraDBClusterRequest, opts ...grpc.CallOption) (*controllerv1beta1.UpdateXtraDBClusterResponse, error) {
-	return c.xtradbClusterClient.UpdateXtraDBCluster(ctx, in, opts...)
+	responseCh := make(chan *apiResponse)
```
Comment (BupycHuk) [inline services/dbaas/dbaas_client.go:306]: I think it's overcomplicated. Please check how API endpoints were disabled for Backup using interceptors. That should be enough.

Comment (BupycHuk) [reply services/dbaas/dbaas_client.go:306]: If required we can use mutexes.

File: services/supervisord/maintail.go [hunk 6]
```diff
@@ -26,13 +26,19 @@ type eventType string
...
-	stopped          eventType = "STOPPED"
-	stopping         eventType = "STOPPING"
-	starting         eventType = "STARTING"
-	running          eventType = "RUNNING"
-	exitedExpected   eventType = "EXITED (expected)"
-	exitedUnexpected eventType = "EXITED (unexpected)"
-	fatal            eventType = "FATAL"
+
+	// Stopping indicates that stop of a process was requested.
+	Stopping eventType = "STOPPING"
```
Comment (BupycHuk) [inline services/supervisord/maintail.go:31]: If we will remove watching supervisord events in dbaasClient we can revert these changes.

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: services/dbaas/dbaas_client.go (hunk 4 above)
Comment (jprukner) [reply services/dbaas/dbaas_client.go]: Done. I use rwmutex to protect the conn to the API.

File: services/dbaas/dbaas_client.go (hunk 5 above)
Comment (jprukner) [reply services/dbaas/dbaas_client.go]: Done, services now have Enabled method.

File: services/supervisord/maintail.go (hunk 6 above)
Comment (jprukner) [reply services/supervisord/maintail.go]: Reverted.

File: services/dbaas/dbaas_client.go [hunk 7]
```diff
@@ -19,102 +19,245 @@ package dbaas
...
+	c.connM.RLock()
+	defer c.connM.RUnlock()
+	c.wg.Add(1)
+
+	in := &controllerv1beta1.CheckKubernetesClusterConnectionRequest{
+		KubeAuth: &controllerv1beta1.KubeAuth{
+			Kubeconfig: kubeConfig,
+		},
+	}
+	out, err := c.kubernetesClient.CheckKubernetesClusterConnection(ctx, in)
```
Comment (BupycHuk) [inline services/dbaas/dbaas_client.go:121]: let's check if connection is exists.

File: services/dbaas/dbaas_client.go [hunk 8]
```diff
@@ -19,102 +19,245 @@ package dbaas
...
 }
 
 // CheckKubernetesClusterConnection checks connection with kubernetes cluster.
 func (c *Client) CheckKubernetesClusterConnection(ctx context.Context, kubeConfig string) (*controllerv1beta1.CheckKubernetesClusterConnectionResponse, error) {
-	return c.kubernetesClient.CheckKubernetesClusterConnection(ctx, &controllerv1beta1.CheckKubernetesClusterConnectionRequest{
-		KubeAuth: &controllerv1beta1.KubeAuth{Kubeconfig: kubeConfig},
-	})
+	c.connM.RLock()
+	defer c.connM.RUnlock()
+	c.wg.Add(1)
```
Comment (BupycHuk) [inline services/dbaas/dbaas_client.go:114]: TBH I don't think that we need WaitGroup here.
RWMutex will not let start `Disconnect` until all `RLocks` are released

File: main.go [hunk 9]
```diff
@@ -732,12 +706,29 @@ func main() {
...
+		if err != nil {
+			l.Fatalf("Failed to connect to dbaas-controller API on %s: %v", *dbaasControllerAPIAddrF, err)
+		}
+		defer func() {
+			err := dbaasClient.Disconnect()
+			if err != nil {
+				l.Fatalf("Failed to disconnect from dbaas-controller API: %v", err)
+			}
+		}()
+	}
```
Comment (BupycHuk) [inline main.go:731]: we can move all this block above https://github.com/percona/pmm-managed/pull/724/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R632

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: services/dbaas/dbaas_client.go [hunk 10]
```diff
@@ -19,102 +19,245 @@ package dbaas
...
+	return resp, err
 }
 
 // CreateXtraDBCluster creates a new XtraDB cluster.
 func (c *Client) CreateXtraDBCluster(ctx context.Context, in *controllerv1beta1.CreateXtraDBClusterRequest, opts ...grpc.CallOption) (*controllerv1beta1.CreateXtraDBClusterResponse, error) {
-	return c.xtradbClusterClient.CreateXtraDBCluster(ctx, in, opts...)
+	c.connM.RLock()
+	defer c.connM.RUnlock()
+	c.wg.Add(1)
+	resp, err := c.xtradbClusterClient.CreateXtraDBCluster(ctx, in, opts...)
```
Comment (palash25) [inline services/dbaas/dbaas_client.go:140]: why are we using waitGroups here? are these function calls async? if not I guess we can remove it

File: main.go (hunk 9 above)
Comment (jprukner) [reply main.go:731]: No, we can't this needs to be after setup is called.

File: services/dbaas/dbaas_client.go (hunk 7 above)
Comment (jprukner) [reply services/dbaas/dbaas_client.go:121]: Do you mean `c.conn`? If it does not exist, it means dbaas is disabled and `Connect` was not called. So I think there is no need to check it.

File: services/dbaas/dbaas_client.go (hunk 10 above)
Comment (jprukner) [reply services/dbaas/dbaas_client.go:133]: WaitGroup was removed.

File: services/dbaas/dbaas_client.go (hunk 8 above)
Comment (jprukner) [reply services/dbaas/dbaas_client.go]: True, I removed WaitGroup.

File: services/dbaas/dbaas_client.go (hunk 7 above)
Comment (jprukner) [reply services/dbaas/dbaas_client.go:117]: We did not do any checking of the result of the rpc call previously, not sure how it would help us now, if that is what you are requesting.

Comment (BupycHuk) [reply services/dbaas/dbaas_client.go:117]: ok

File: main.go (hunk 9 above)
Comment (BupycHuk) [reply main.go:731]: Why?

Comment (BupycHuk) [reply main.go:731]: ok, I see. then we can just move https://github.com/percona/pmm-managed/pull/724/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R715 above and keep other lines here.

**Review by palash25 — APPROVED** [review]

File: main.go (hunk 9 above)
Comment (jprukner) [reply main.go:731]: Ok, good point.

File: services/dbaas/dbaas_client.go [hunk 11]
```diff
@@ -19,102 +19,213 @@ package dbaas
...
-	})
+	c.connM.RLock()
+	defer c.connM.RUnlock()
+	in := &controllerv1beta1.CheckKubernetesClusterConnectionRequest{
+		KubeAuth: &controllerv1beta1.KubeAuth{
+			Kubeconfig: kubeConfig,
+		},
+	}
+	out, err := c.kubernetesClient.CheckKubernetesClusterConnection(ctx, in)
+	out.GetStatus()
```
Comment (BupycHuk) [inline services/dbaas/dbaas_client.go:118]: this can be removed

File: services/dbaas/dbaas_client.go [hunk 12]
```diff
@@ -19,102 +19,213 @@ package dbaas
...
+	return out, err
```
Comment (BupycHuk) [inline services/dbaas/dbaas_client.go:119]:
```suggestion
	return c.kubernetesClient.CheckKubernetesClusterConnection(ctx, in)
```

File: services/dbaas/dbaas_client_test.go [hunk 13]
```diff
@@ -19,52 +19,58 @@ package dbaas
...
 	"testing"
 	"time"
 
-	"github.com/percona/pmm/version"
 	"github.com/stretchr/testify/require"
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/backoff"
 )
 
+const supervisordConfigDir = "/etc/supervisord.d"
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply services/dbaas/dbaas_client_test.go:29]: looks like it's actual problem

File: services/dbaas/dbaas_client_test.go [hunk 14]
```diff
@@ -19,52 +19,58 @@ package dbaas
...
-		if os.Getenv("PERCONA_TEST_DBAAS") != "1" { // FIXME use strconv.ParseBool
-			t.Skip("PERCONA_TEST_DBAAS env variable is not passed, skipping")
+	t.Run("ValidKubeConfig", func(t *testing.T) {
+		v := os.Getenv("ENABLE_DBAAS")
+		dbaasEnabled, err := strconv.ParseBool(v)
+		if err != nil {
+			t.Skipf("Invalid value %q for environment variable ENABLE_DBAAS", v)
+		}
+		if !dbaasEnabled {
+			t.Skip("DBaaS is not enabled")
```
Comment (BupycHuk) [inline services/dbaas/dbaas_client_test.go:51]: we can move it to parent test beginning

File: services/management/dbaas/version_service_client.go [hunk 15]
```diff
@@ -25,10 +25,10 @@ import (
 	"strings"
 	"time"
 
+	"github.com/percona/pmm-managed/utils/irt"
```
Comment (BupycHuk) [inline services/management/dbaas/version_service_client.go:28]: it should be reverted

File: services/dbaas/dbaas_client.go (hunk 11 above)
Comment (jprukner) [reply services/dbaas/dbaas_client.go]: uh, removed

File: services/dbaas/dbaas_client.go (hunk 12 above)
Comment (jprukner) [reply services/dbaas/dbaas_client.go]: done in all other calls

File: services/dbaas/dbaas_client_test.go (hunk 14 above)
Comment (jprukner) [reply services/dbaas/dbaas_client_test.go]: done

File: services/management/dbaas/version_service_client.go (hunk 15 above)
Comment (jprukner) [reply services/management/dbaas/version_service_client.go]: reverted

**Review by BupycHuk — APPROVED** [review]

File: services/victoriametrics/prometheus.go [hunk 16]
```diff
@@ -204,7 +204,7 @@ func AddInternalServicesToScrape(cfg *config.Config, s models.MetricsResolutions
 		scrapeConfigForPMMManaged(s.MR),
 		scrapeConfigForQANAPI2(s.MR),
 	)
-	// TODO Refactor to remove boolean positional parameter when Prometheus or PERCONA_TEST_DBAAS is removed
+	// TODO Refactor to remove boolean positional parameter when Prometheus is removed
 	if dbaas {
```
Comment (BupycHuk) [inline services/victoriametrics/prometheus.go:208]: Do we regenerate scrape config on enable/disable DBaaS?

**Review by BupycHuk — APPROVED** [review]

File: services/victoriametrics/prometheus.go (hunk 16 above)
Comment (jprukner) [reply services/victoriametrics/prometheus.go:208]: Yes, we do. Does it affect something else than dbaas-controller? Only difference is we scrape dbaas-controller only if it's running.

Comment (jprukner) [reply services/victoriametrics/prometheus.go:208]: Yes, we do as part of server.UpdateConfigurations method.
